### PR TITLE
Locking appropriately before closing the channel to indicate migration

### LINKED ISSUE #2230: Host panic - nomad 0.5.2 - client.runAllocs
Host panic, and nomad was unable to restart afterwords.

```
.....
    2017/01/23 17:30:12.568052 [DEBUG] client: allocs: (added 4) (removed 0) (updated 22) (ignore 18)
    2017/01/23 17:30:12.568193 [DEBUG] client: added alloc "c54717a6-7b63-428f-d569-a8a3553907d2" to blocked queue for previous allocation "c294bf88-8afa-5e6a-e123-14a671d77fa5"
    2017/01/23 17:30:12.584264 [DEBUG] client: allocs: (added 4) (removed 0) (updated 22) (ignore 18)
panic: close of closed channel

goroutine 131 [running]:
panic(0xfc0740, 0xc45c60b5f0)
        /opt/go/src/runtime/panic.go:500 +0x1a1
github.com/hashicorp/nomad/client.(*Client).runAllocs(0xc420080dc0, 0xc48f1ca880)
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:1369 +0xde0
github.com/hashicorp/nomad/client.(*Client).run(0xc420080dc0)
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:927 +0x119
created by github.com/hashicorp/nomad/client.NewClient
        /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:266 +0xc78
    Loaded configuration from /etc/nomad/config.hcl
==> Starting Nomad agent...
==> Error starting agent: client setup failed: failed to restore state: 1 error(s) occurred:

* 1 error(s) occurred:

* alloc_runner snapshot includes a nil context
    Loaded configuration from /etc/nomad/config.hcl

```
### Nomad version
Output from `nomad version`
Nomad v0.5.2

### Operating system and Environment details
Linux ip-10-XXX-X-XXX 4.4.0-59-generic #80-Ubuntu SMP Fri Jan 6 17:47:47 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux

### Issue
Runtime panic. Developer was using the logs API endpoint heavily during this time.

### Reproduction steps
Not sure -it was otherwise normal usage. Heavy use of logs, however.

### Nomad Client logs (if appropriate)
[client.zip](https://github.com/hashicorp/nomad/files/724492/client.zip)

### PR DESCRIPTION
**Pull request by diptanu**
Fixes #2230

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -150,14 +150,37 @@ type Client struct {
...
 
 	// garbageCollector is used to garbage collect terminal allocations present
 	// in the node automatically
 	garbageCollector *AllocGarbageCollector
 }
 
+// migrateAllocCtrl indicates whether migration is complete
+type migrateAllocCtrl struct {
+	ch     chan struct{}
+	chLock sync.Mutex
```
Comment (dadgar) [inline client/client.go]: Add `closed bool`

File: client/client.go [hunk 2]
```diff
@@ -150,14 +150,37 @@ type Client struct {
...
+	return &migrateAllocCtrl{
+		ch: make(chan struct{}),
+	}
+}
+
+func (m *migrateAllocCtrl) closeCh() {
+	m.chLock.Lock()
+	defer m.chLock.Unlock()
+
+	// If channel is not closed then close it
```
Comment (dadgar) [inline client/client.go]:
```
m.chLock.Lock()
defer m.chLock.Unlock()

if m.closed {
   return
}

m.closed = true
close(m.ch)
```

Comment (dadgar) [inline client/client.go]: The current implementation will actually block

Comment (diptanu) [inline client/client.go]: Yeah it doesn't block and return false immediately when the ch is closed.

File: client/client.go [hunk 3]
```diff
@@ -150,14 +150,37 @@ type Client struct {
...
+}
+
+func newMigrateAllocCtrl() *migrateAllocCtrl {
+	return &migrateAllocCtrl{
+		ch: make(chan struct{}),
+	}
+}
+
+func (m *migrateAllocCtrl) closeCh() {
+	m.chLock.Lock()
```
Comment (schmichael) [inline client/client.go]: I don't think you need this lock as `runAllocs` method is only ever called from a single goroutine so all chan mutations are serialized.

Comment (diptanu) [inline client/client.go]: It might be called concurrently though.

**Comment by dadgar** [comment]
@diptanu Broke a test https://travis-ci.org/hashicorp/nomad/builds/194574408#L915

**Pull request merged**